src/target/breakpoints.c

```diff
@@ -287,60 +295,89 @@ static void breakpoint_free(struct target *target, struct breakpoint *breakpoint
 	if (!breakpoint)
 		return;
 
-	retval = target_remove_breakpoint(target, breakpoint);
+	retval = target_remove_breakpoint(breakpoint_target, breakpoint);
 
 	LOG_DEBUG("free BPID: %" PRIu32 " --> %d", breakpoint->unique_id, retval);
 	(*breakpoint_p) = breakpoint->next;
 	free(breakpoint->orig_instr);
 	free(breakpoint);
 }
 
-static int breakpoint_remove_internal(struct target *target, target_addr_t address)
-{
-	struct breakpoint *breakpoint = target->breakpoints;
-
-	while (breakpoint) {
-		if ((breakpoint->address == address) ||
-		    (breakpoint->address == 0 && breakpoint->asid == address))
-			break;
-		breakpoint = breakpoint->next;
-	}
-
-	if (breakpoint) {
-		breakpoint_free(target, breakpoint);
-		return 1;
-	} else {
-		if (!target->smp)
-			LOG_ERROR("no breakpoint at address " TARGET_ADDR_FMT " found", address);
-		return 0;
-	}
-}
-
 static void breakpoint_remove_all_internal(struct target *target)
 {
 	struct breakpoint *breakpoint = target->breakpoints;
 
 	while (breakpoint) {
 		struct breakpoint *tmp = breakpoint;
 		breakpoint = breakpoint->next;
-		breakpoint_free(target, tmp);
+		breakpoint_free(target, target, tmp);
 	}
 }
 
 void breakpoint_remove(struct target *target, target_addr_t address)
 {
 	if (target->smp) {
-		unsigned int num_breakpoints = 0;
+		unsigned int found = 0;
 		struct target_list *head;
+		/* Target where we found a software breakpoint. */
+		struct target *software_breakpoint_target = NULL;
+		struct breakpoint *software_breakpoint = NULL;
+		/* Target that is available. */
+		struct target *available_target = NULL;
+		/* Target that is available and halted. */
+		struct target *halted_target = NULL;
 
 		foreach_smp_target(head, target->smp_targets) {
 			struct target *curr = head->target;
-			num_breakpoints += breakpoint_remove_internal(curr, address);
+
+			if (!available_target && curr->state != TARGET_UNAVAILABLE)
+				available_target = curr;
+			if (!halted_target && curr->state == TARGET_HALTED)
+				halted_target = curr;
+
+			struct breakpoint *breakpoint = breakpoint_find(curr, address);
+			if (!breakpoint)
+				continue;
+
+			found++;
+
+			if (breakpoint->type == BKPT_SOFT) {
+				/* Software breakpoints are set on only one of
+				 * the SMP targets.  We can remove them through
+				 * any of the SMP targets. */
+				assert(!software_breakpoint_target);
+				software_breakpoint_target = curr;
+				software_breakpoint = breakpoint;
+				continue;
```

Comment:
I like the !target->smp check, but I don't think the other refactor works. We don't have a breakpoint object until we call breakpoint_find(). Each target might have different breakpoints set on the same address. I don't think you're supposed to be able to get into that state, but it seems better to be a bit defensive about that. I prefer to keep the single loop, which will delete any hardware breakpoints it finds, and also keeps track of available target, halted target, and software breakpoint target.

Reply:
>  I don't think the other refactor works ... We don't have a breakpoint object until we call breakpoint_find() ...

Ah, I missed that we don't know the breakpoint type upfront (only the address). In that case, let's leave it as is.

Another thing is that the current algorithm assumes that the SW breakpoint will be found on a single target only. To stay defensive, would it make sense to warn if that is not the case (instead of the assert)?

Reply:
Yes. That makes sense. Especially with the `smp` commands that let you modify SMP groups at run time, you could end up in a situation where this assert is fired due to user configuration.